tools/c7n_gcp/c7n_gcp/query.py

```diff
@@ -145,7 +145,45 @@ def get_resource_query(self):
             return {'filter': self.data.get('query')}
 
     def resources(self, query=None):
-        q = query or self.get_resource_query()
+        if hasattr(self.resource_type, 'parent'):
+            return self._fetch_resources_relationally(query)
+
+        return self._fetch_resources_directly(query)
+
+    def _fetch_resources_relationally(self, query):
+        """ Fetches resources by first fetching their parents and extracting
+        required arguments from the latter ones. Injects each corresponding
+        parent into a child resource afterwards.
+        """
+
+        resources = []
+
+        parent_description = self.resource_type.parent
+        resource_type, args_extractor = parent_description[:-1]
```

Comment:
this doesn't work if the third param is truly optional. rather then using a tuple i suggest making this an actual dictionary

```python
>>> x = (1,2,3)
>>> y = (1, 2)
>>> a, b = x[:-1]                                                                                                                                                     
>>> a, b
(1, 2)
>>> a, b = y[:-1]                                                                                                                                                     
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: need more than 1 value to unpack
```